File: ApiService.kt
import retrofit2.Response
import retrofit2.http.*
import okhttp3.MultipartBody
import okhttp3.RequestBody

data class AnalyticsSummaryResponse(
    val user_count: Int,
    val family_count: Int,
    val task_count: Int,
    val match_count: Int,
    val background_check_count: Int
)

data class RoleBreakdownResponse(
    val role: String,
    val count: Int
)

data class NotificationEntity(
    val id: Int,
    val user_id: Int?,
    val message: String,
    val sent_at: String,
    val is_read: Boolean
)

data class UnreadCountResponse(
    val unread: Int
)

interface ApiService {
    // USERS
    @GET("users")
    suspend fun getUsers(@Header("Authorization") token: String): List<UsersEntity>
    @POST("users")
    suspend fun createUser(@Header("Authorization") token: String, @Body user: UsersEntity): Response<Unit>
    @PUT("users/{id}")
    suspend fun updateUser(@Header("Authorization") token: String, @Path("id") id: Int, @Body user: UsersEntity): Response<Unit>
    @DELETE("users/{id}")
    suspend fun deleteUser(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // CHILDREN
    @GET("children")
    suspend fun getChildren(@Header("Authorization") token: String): List<ChildrenEntity>
    @POST("children")
    suspend fun createChild(@Header("Authorization") token: String, @Body child: ChildrenEntity): Response<Unit>
    @PUT("children/{id}")
    suspend fun updateChild(@Header("Authorization") token: String, @Path("id") id: Int, @Body child: ChildrenEntity): Response<Unit>
    @DELETE("children/{id}")
    suspend fun deleteChild(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // GUARDIANS
    @GET("guardians")
    suspend fun getGuardians(@Header("Authorization") token: String): List<GuardiansEntity>
    @POST("guardians")
    suspend fun createGuardian(@Header("Authorization") token: String, @Body guardian: GuardiansEntity): Response<Unit>
    @PUT("guardians/{id}")
    suspend fun updateGuardian(@Header("Authorization") token: String, @Path("id") id: Int, @Body guardian: GuardiansEntity): Response<Unit>
    @DELETE("guardians/{id}")
    suspend fun deleteGuardian(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // COURT CASES
    @GET("court_cases")
    suspend fun getCourtCases(@Header("Authorization") token: String): List<CourtCasesEntity>
    @POST("court_cases")
    suspend fun createCourtCase(@Header("Authorization") token: String, @Body case: CourtCasesEntity): Response<Unit>
    @PUT("court_cases/{id}")
    suspend fun updateCourtCase(@Header("Authorization") token: String, @Path("id") id: Int, @Body case: CourtCasesEntity): Response<Unit>
    @DELETE("court_cases/{id}")
    suspend fun deleteCourtCase(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // PLACEMENTS
    @GET("placements")
    suspend fun getPlacements(@Header("Authorization") token: String): List<PlacementsEntity>
    @POST("placements")
    suspend fun createPlacement(@Header("Authorization") token: String, @Body placement: PlacementsEntity): Response<Unit>
    @PUT("placements/{id}")
    suspend fun updatePlacement(@Header("Authorization") token: String, @Path("id") id: Int, @Body placement: PlacementsEntity): Response<Unit>
    @DELETE("placements/{id}")
    suspend fun deletePlacement(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // MEDICAL RECORDS
    @GET("medical_records")
    suspend fun getMedicalRecords(@Header("Authorization") token: String): List<MedicalRecordsEntity>
    @POST("medical_records")
    suspend fun createMedicalRecord(@Header("Authorization") token: String, @Body record: MedicalRecordsEntity): Response<Unit>
    @PUT("medical_records/{id}")
    suspend fun updateMedicalRecord(@Header("Authorization") token: String, @Path("id") id: Int, @Body record: MedicalRecordsEntity): Response<Unit>
    @DELETE("medical_records/{id}")
    suspend fun deleteMedicalRecord(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // CASE REPORTS
    @GET("case_reports")
    suspend fun getCaseReports(@Header("Authorization") token: String): List<CaseReportsEntity>
    @POST("case_reports")
    suspend fun createCaseReport(@Header("Authorization") token: String, @Body report: CaseReportsEntity): Response<Unit>
    @PUT("case_reports/{id}")
    suspend fun updateCaseReport(@Header("Authorization") token: String, @Path("id") id: Int, @Body report: CaseReportsEntity): Response<Unit>
    @DELETE("case_reports/{id}")
    suspend fun deleteCaseReport(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // MONEY RECORDS
    @GET("money_records")
    suspend fun getMoneyRecords(@Header("Authorization") token: String): List<MoneyRecordsEntity>
    @POST("money_records")
    suspend fun createMoneyRecord(@Header("Authorization") token: String, @Body record: MoneyRecordsEntity): Response<Unit>
    @PUT("money_records/{id}")
    suspend fun updateMoneyRecord(@Header("Authorization") token: String, @Path("id") id: Int, @Body record: MoneyRecordsEntity): Response<Unit>
    @DELETE("money_records/{id}")
    suspend fun deleteMoneyRecord(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // EDUCATION RECORDS
    @GET("education_records")
    suspend fun getEducationRecords(@Header("Authorization") token: String): List<EducationRecordsEntity>
    @POST("education_records")
    suspend fun createEducationRecord(@Header("Authorization") token: String, @Body record: EducationRecordsEntity): Response<Unit>
    @PUT("education_records/{id}")
    suspend fun updateEducationRecord(@Header("Authorization") token: String, @Path("id") id: Int, @Body record: EducationRecordsEntity): Response<Unit>
    @DELETE("education_records/{id}")
    suspend fun deleteEducationRecord(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // DOCUMENTS
    @GET("documents")
    suspend fun getDocuments(@Header("Authorization") token: String): List<DocumentsEntity>
    @POST("documents")
    suspend fun createDocument(@Header("Authorization") token: String, @Body document: DocumentsEntity): Response<Unit>
    @PUT("documents/{id}")
    suspend fun updateDocument(@Header("Authorization") token: String, @Path("id") id: Int, @Body document: DocumentsEntity): Response<Unit>
    @DELETE("documents/{id}")
    suspend fun deleteDocument(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // AUDIT LOGS
    @GET("audit_logs")
    suspend fun getAuditLogs(@Header("Authorization") token: String): List<AuditLogsEntity>
    @POST("audit_logs")
    suspend fun createAuditLog(@Header("Authorization") token: String, @Body log: AuditLogsEntity): Response<Unit>
    @PUT("audit_logs/{id}")
    suspend fun updateAuditLog(@Header("Authorization") token: String, @Path("id") id: Int, @Body log: AuditLogsEntity): Response<Unit>
    @DELETE("audit_logs/{id}")
    suspend fun deleteAuditLog(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // PERMISSIONS
    @GET("permissions")
    suspend fun getPermissions(@Header("Authorization") token: String): List<PermissionsEntity>
    @POST("permissions")
    suspend fun createPermission(@Header("Authorization") token: String, @Body permission: PermissionsEntity): Response<Unit>
    @PUT("permissions/{id}")
    suspend fun updatePermission(@Header("Authorization") token: String, @Path("id") id: Int, @Body permission: PermissionsEntity): Response<Unit>
    @DELETE("permissions/{id}")
    suspend fun deletePermission(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // USER PERMISSIONS
    @GET("user_permissions")
    suspend fun getUserPermissions(@Header("Authorization") token: String): List<UserPermissionsEntity>
    @POST("user_permissions")
    suspend fun createUserPermission(@Header("Authorization") token: String, @Body userPermission: UserPermissionsEntity): Response<Unit>
    @PUT("user_permissions/{user_id}/{permission_id}")
    suspend fun updateUserPermission(@Header("Authorization") token: String, @Path("user_id") userId: Int, @Path("permission_id") permissionId: Int, @Body userPermission: UserPermissionsEntity): Response<Unit>
    @DELETE("user_permissions/{user_id}/{permission_id}")
    suspend fun deleteUserPermission(@Header("Authorization") token: String, @Path("user_id") userId: Int, @Path("permission_id") permissionId: Int): Response<Unit>

    // FAMILY PROFILE
    @GET("family_profile")
    suspend fun getFamilyProfiles(@Header("Authorization") token: String): List<FamilyProfileEntity>
    @POST("family_profile")
    suspend fun createFamilyProfile(@Header("Authorization") token: String, @Body profile: FamilyProfileEntity): Response<Unit>
    @PUT("family_profile/{id}")
    suspend fun updateFamilyProfile(@Header("Authorization") token: String, @Path("id") id: Int, @Body profile: FamilyProfileEntity): Response<Unit>
    @DELETE("family_profile/{id}")
    suspend fun deleteFamilyProfile(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // FOSTER TASKS
    @GET("foster_tasks")
    suspend fun getFosterTasks(@Header("Authorization") token: String): List<FosterTasksEntity>
    @POST("foster_tasks")
    suspend fun createFosterTask(@Header("Authorization") token: String, @Body task: FosterTasksEntity): Response<Unit>
    @PUT("foster_tasks/{id}")
    suspend fun updateFosterTask(@Header("Authorization") token: String, @Path("id") id: Int, @Body task: FosterTasksEntity): Response<Unit>
    @DELETE("foster_tasks/{id}")
    suspend fun deleteFosterTask(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // FOSTER MATCHES
    @GET("foster_matches")
    suspend fun getFosterMatches(@Header("Authorization") token: String): List<FosterMatchesEntity>
    @POST("foster_matches")
    suspend fun createFosterMatch(@Header("Authorization") token: String, @Body match: FosterMatchesEntity): Response<Unit>
    @PUT("foster_matches/{id}")
    suspend fun updateFosterMatch(@Header("Authorization") token: String, @Path("id") id: Int, @Body match: FosterMatchesEntity): Response<Unit>
    @DELETE("foster_matches/{id}")
    suspend fun deleteFosterMatch(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // BACKGROUND CHECKS
    @GET("background_checks")
    suspend fun getBackgroundChecks(@Header("Authorization") token: String): List<BackgroundChecksEntity>
    @POST("background_checks")
    suspend fun createBackgroundCheck(@Header("Authorization") token: String, @Body check: BackgroundChecksEntity): Response<Unit>
    @PUT("background_checks/{id}")
    suspend fun updateBackgroundCheck(@Header("Authorization") token: String, @Path("id") id: Int, @Body check: BackgroundChecksEntity): Response<Unit>
    @DELETE("background_checks/{id}")
    suspend fun deleteBackgroundCheck(@Header("Authorization") token: String, @Path("id") id: Int): Response<Unit>

    // --- ANALYTICS ---
    @GET("analytics/summary")
    suspend fun getAnalyticsSummary(@Header("Authorization") token: String): AnalyticsSummaryResponse

    @GET("analytics/roles")
    suspend fun getRoleBreakdown(@Header("Authorization") token: String): List<RoleBreakdownResponse>

    @GET("analytics/recent-activity")
    suspend fun getRecentActivity(@Header("Authorization") token: String): List<AuditLogsEntity>

    @GET("analytics/pending-background-checks")
    suspend fun getPendingBackgroundChecks(@Header("Authorization") token: String): List<BackgroundChecksEntity>

    // --- NOTIFICATIONS ---
    @GET("notifications/history")
    suspend fun getNotificationHistory(@Header("Authorization") token: String): List<NotificationEntity>

    @GET("notifications/unread-count")
    suspend fun getUnreadNotificationCount(@Header("Authorization") token: String): UnreadCountResponse

    // --- FILE/PHOTO UPLOADS ---
    @Multipart
    @POST("children/{id}/photo")
    suspend fun uploadChildPhoto(
        @Header("Authorization") token: String,
        @Path("id") childId: Int,
        @Part photo: MultipartBody.Part
    ): Response<Unit>

    @Multipart
    @POST("documents/{id}/file")
    suspend fun uploadDocumentFile(
        @Header("Authorization") token: String,
        @Path("id") documentId: Int,
        @Part file: MultipartBody.Part
    ): Response<Unit>
} 

File: src/js/api-service.js
// API Service for Adoption & Child Care System
const API_BASE_URL = 'http://192.168.43.197:50000';

class ApiService {
    constructor() {
        this.baseURL = API_BASE_URL;
        this.token = localStorage.getItem('authToken');
    }

    setToken(token) {
        this.token = token;
        localStorage.setItem('authToken', token);
    }

    clearToken() {
        this.token = null;
        localStorage.removeItem('authToken');
    }

    getHeaders() {
        const headers = {
            'Content-Type': 'application/json'
        };
        if (this.token) {
            headers['Authorization'] = `Bearer ${this.token}`;
        }
        return headers;
    }

    async request(endpoint, options = {}) {
        const url = `${this.baseURL}${endpoint}`;
        const config = {
            ...options,
            headers: {
                ...this.getHeaders(),
                ...options.headers
            }
        };

        try {
            const response = await fetch(url, config);
            const data = await response.json();
            
            if (!response.ok) {
                throw new Error(data.message || 'API request failed');
            }
            
            return data;
        } catch (error) {
            console.error('API Error:', error);
            throw error;
        }
    }

    // Generic CRUD operations for any table
    async getAll(table) {
        return this.request(`/${table}`);
    }

    async getById(table, id) {
        return this.request(`/${table}/${id}`);
    }

    async create(table, data) {
        return this.request(`/${table}`, {
            method: 'POST',
            body: JSON.stringify(data)
        });
    }

    async update(table, id, data) {
        return this.request(`/${table}/${id}`, {
            method: 'PUT',
            body: JSON.stringify(data)
        });
    }

    async delete(table, id) {
        return this.request(`/${table}/${id}`, {
            method: 'DELETE'
        });
    }

    // Authentication
    async login(email, password) {
        const response = await this.request('/login', {
            method: 'POST',
            body: JSON.stringify({ email, password })
        });
        
        if (response.token) {
            this.setToken(response.token);
        }
        
        return response;
    }

    async register(userData) {
        return this.request('/register', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify(userData)
        });
    }

    // Specific endpoints for main entities
    
    // Children
    async getChildren() {
        return this.getAll('children');
    }

    async getChildById(id) {
        return this.getById('children', id);
    }

    async updateChild(id, data) {
        return this.update('children', id, data);
    }

    async deleteChild(id) {
        return this.delete('children', id);
    }

    // Users
    async getUsers() {
        return this.getAll('users');
    }

    async getUserById(id) {
        return this.getById('users', id);
    }

    async updateUser(id, data) {
        return this.update('users', id, data);
    }

    async deleteUser(id) {
        return this.delete('users', id);
    }

    // Families
    async getFamilies() {
        return this.getAll('families');
    }

    async getFamilyProfile() {
        return this.request('/family_profile');
    }

    async createFamilyProfile(data) {
        return this.request('/family_profile', {
            method: 'POST',
            body: JSON.stringify(data)
        });
    }

    async updateFamilyProfile(id, data) {
        return this.request(`/family_profile/${id}`, {
            method: 'PUT',
            body: JSON.stringify(data)
        });
    }

    async deleteFamilyProfile(id) {
        return this.request(`/family_profile/${id}`, {
            method: 'DELETE'
        });
    }

    // Guardians
    async getGuardians() {
        return this.getAll('guardians');
    }

    async updateGuardian(id, data) {
        return this.update('guardians', id, data);
    }

    async deleteGuardian(id) {
        return this.delete('guardians', id);
    }

    // Placements
    async getPlacements() {
        return this.getAll('placements');
    }

    async updatePlacement(id, data) {
        return this.update('placements', id, data);
    }

    async deletePlacement(id) {
        return this.delete('placements', id);
    }

    // Adoption Applications
    async getAdoptionApplications() {
        return this.getAll('adoption_applications');
    }

    async createAdoptionApplication(data) {
        return this.create('adoption_applications', data);
    }

    async updateAdoptionApplication(id, data) {
        return this.update('adoption_applications', id, data);
    }

    async deleteAdoptionApplication(id) {
        return this.delete('adoption_applications', id);
    }

    // Foster Care
    async getFosterTasks() {
        return this.request('/foster_tasks');
    }

    async createFosterTask(data) {
        return this.request('/foster_tasks', {
            method: 'POST',
            body: JSON.stringify(data)
        });
    }

    async updateFosterTask(id, data) {
        return this.request(`/foster_tasks/${id}`, {
            method: 'PUT',
            body: JSON.stringify(data)
        });
    }

    async deleteFosterTask(id) {
        return this.request(`/foster_tasks/${id}`, {
            method: 'DELETE'
        });
    }

    async getFosterMatches() {
        return this.request('/foster_matches');
    }

    async createFosterMatch(data) {
        return this.request('/foster_matches', {
            method: 'POST',
            body: JSON.stringify(data)
        });
    }

    async updateFosterMatch(id, data) {
        return this.request(`/foster_matches/${id}`, {
            method: 'PUT',
            body: JSON.stringify(data)
        });
    }

    async deleteFosterMatch(id) {
        return this.request(`/foster_matches/${id}`, {
            method: 'DELETE'
        });
    }

    // Background Checks
    async getBackgroundChecks() {
        return this.request('/background_checks');
    }

    async createBackgroundCheck(data) {
        return this.request('/background_checks', {
            method: 'POST',
            body: JSON.stringify(data)
        });
    }

    async updateBackgroundCheck(id, data) {
        return this.request(`/background_checks/${id}`, {
            method: 'PUT',
            body: JSON.stringify(data)
        });
    }

    async deleteBackgroundCheck(id) {
        return this.request(`/background_checks/${id}`, {
            method: 'DELETE'
        });
    }

    async triggerBackgroundCheck(userId) {
        return this.request(`/background_checks/${userId}/trigger`, {
            method: 'POST'
        });
    }

    async getBackgroundCheckForUser(userId) {
        return this.request(`/background_checks/${userId}`);
    }

    // Medical Records
    async getMedicalRecords() {
        return this.getAll('medical_records');
    }

    async updateMedicalRecord(id, data) {
        return this.update('medical_records', id, data);
    }

    async deleteMedicalRecord(id) {
        return this.delete('medical_records', id);
    }

    // Education Records
    async getEducationRecords() {
        return this.getAll('education_records');
    }

    async updateEducationRecord(id, data) {
        return this.update('education_records', id, data);
    }

    async deleteEducationRecord(id) {
        return this.delete('education_records', id);
    }

    // Documents
    async getDocuments() {
        return this.getAll('documents');
    }

    async updateDocument(id, data) {
        return this.update('documents', id, data);
    }

    async deleteDocument(id) {
        return this.delete('documents', id);
    }

    // Case Reports
    async getCaseReports() {
        return this.getAll('case_reports');
    }

    async updateCaseReport(id, data) {
        return this.update('case_reports', id, data);
    }

    async deleteCaseReport(id) {
        return this.delete('case_reports', id);
    }

    // Court Cases
    async getCourtCases() {
        return this.getAll('court_cases');
    }

    async updateCourtCase(id, data) {
        return this.update('court_cases', id, data);
    }

    async deleteCourtCase(id) {
        return this.delete('court_cases', id);
    }

    // Home Studies
    async getHomeStudies() {
        return this.getAll('home_studies');
    }

    async createHomeStudy(data) {
        return this.create('home_studies', data);
    }

    async updateHomeStudy(id, data) {
        return this.update('home_studies', id, data);
    }

    async deleteHomeStudy(id) {
        return this.delete('home_studies', id);
    }

    // Money Records
    async getMoneyRecords() {
        return this.getAll('money_records');
    }

    async updateMoneyRecord(id, data) {
        return this.update('money_records', id, data);
    }

    async deleteMoneyRecord(id) {
        return this.delete('money_records', id);
    }

    // Notifications
    async getNotifications() {
        return this.request('/notifications');
    }

    async createNotification(data) {
        return this.request('/notifications', {
            method: 'POST',
            body: JSON.stringify(data)
        });
    }

    async markNotificationAsRead(id) {
        return this.request(`/notifications/${id}/read`, {
            method: 'PUT'
        });
    }

    async getUnreadNotificationCount() {
        return this.request('/notifications/unread-count');
    }

    async sendNotification(data) {
        return this.request('/notifications/send', {
            method: 'POST',
            body: JSON.stringify(data)
        });
    }

    // Analytics
    async getAnalyticsSummary() {
        return this.request('/analytics/summary');
    }

    async getRoleAnalytics() {
        return this.request('/analytics/roles');
    }

    async getRecentActivity() {
        return this.request('/analytics/recent-activity');
    }

    async getPendingBackgroundChecks() {
        return this.request('/analytics/pending-background-checks');
    }

    // Audit Logs
    async getAuditLogs() {
        return this.getAll('audit_logs');
    }

    async updateAuditLog(id, data) {
        return this.update('audit_logs', id, data);
    }

    async deleteAuditLog(id) {
        return this.delete('audit_logs', id);
    }

    // Permissions
    async getPermissions() {
        return this.getAll('permissions');
    }

    async updatePermission(id, data) {
        return this.update('permissions', id, data);
    }

    async deletePermission(id) {
        return this.delete('permissions', id);
    }

    // User Permissions
    async updateUserPermission(userId, permissionId, data) {
        return this.request(`/user_permissions/${userId}/${permissionId}`, {
            method: 'PUT',
            body: JSON.stringify(data)
        });
    }

    async deleteUserPermission(userId, permissionId) {
        return this.request(`/user_permissions/${userId}/${permissionId}`, {
            method: 'DELETE'
        });
    }

    // Location Services
    async updateUserLocation(id, latitude, longitude) {
        return this.request(`/users/${id}/location`, {
            method: 'PUT',
            body: JSON.stringify({ latitude, longitude })
        });
    }

    async getAllUserLocations() {
        return this.request('/users/locations');
    }

    async updateFamilyLocation(id, latitude, longitude) {
        return this.request(`/family_profile/${id}/location`, {
            method: 'PUT',
            body: JSON.stringify({ latitude, longitude })
        });
    }

    async getAllFamilyLocations() {
        return this.request('/family_profile/locations');
    }

    // Matching System
    async findMatches(criteria) {
        return this.request('/match', {
            method: 'POST',
            body: JSON.stringify(criteria)
        });
    }

    // Photo Upload
    async uploadPhoto(formData) {
        return this.request('/upload-photo', {
            method: 'POST',
            headers: {}, // Let browser set Content-Type for FormData
            body: formData
        });
    }

    async updateUserPhoto(id, formData) {
        return this.request(`/users/${id}/photo`, {
            method: 'PUT',
            headers: {}, // Let browser set Content-Type for FormData
            body: formData
        });
    }

    // Emergency Services
    async sendEmergencyAlert(data) {
        return this.request('/api/v2/emergency/alert', {
            method: 'POST',
            body: JSON.stringify(data)
        });
    }

    async sendEmergencyLocation(data) {
        return this.request('/api/v2/emergency/location', {
            method: 'POST',
            body: JSON.stringify(data)
        });
    }

    // Sync Services
    async syncPull() {
        return this.request('/api/v2/sync/pull');
    }

    async syncPush(data) {
        return this.request('/api/v2/sync/push', {
            method: 'POST',
            body: JSON.stringify(data)
        });
    }

    // FCM Tokens
    async registerFCMToken(token) {
        return this.request('/api/fcm/token', {
            method: 'POST',
            body: JSON.stringify({ token })
        });
    }

    // Utility
    async checkUserExists(email) {
        return this.request('/user-exists', {
            method: 'POST',
            body: JSON.stringify({ email })
        });
    }

    async getDatabaseStatus() {
        return this.request('/admin-db-status');
    }
}

// Create singleton instance
const apiService = new ApiService();

// Make it available globally
window.apiService = apiService;